File: JAVA/java/src/BalanceaBinarySearchTree.java
import java.util.*;

// Definition for a binary tree node.

public class BalanceaBinarySearchTree {

    // Method to balance a BST
    public TreeNode balanceBST(TreeNode root) {
        // Step 1: Perform in-order traversal to collect all node values into a sorted list
        List<Integer> sortedList = new ArrayList<>();
        inorderTraversal(root, sortedList);

        // Step 2: Construct a balanced BST from sorted list
        return constructBalancedBST(sortedList, 0, sortedList.size() - 1);
    }

    // In-order traversal to collect node values into sorted list
    private void inorderTraversal(TreeNode node, List<Integer> sortedList) {
        if (node == null) {
            return;
        }
        inorderTraversal(node.left, sortedList);
        sortedList.add(node.val);
        inorderTraversal(node.right, sortedList);
    }

    // Recursive method to construct a balanced BST from sorted list
    private TreeNode constructBalancedBST(List<Integer> sortedList, int start, int end) {
        if (start > end) {
            return null;
        }
        int mid = (start + end) / 2;
        TreeNode root = new TreeNode(sortedList.get(mid));
        root.left = constructBalancedBST(sortedList, start, mid - 1);
        root.right = constructBalancedBST(sortedList, mid + 1, end);
        return root;
    }

    // Method to convert balanced BST to level-order list format
    public List<Integer> bstToLevelOrder(TreeNode root) {
        List<Integer> levelOrderList = new ArrayList<>();
        if (root == null) {
            return levelOrderList;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();
            if (current != null) {
                levelOrderList.add(current.val);
                queue.offer(current.left);
                queue.offer(current.right);
            } else {
                levelOrderList.add(null);
            }
        }

        // Remove trailing nulls
        while (!levelOrderList.isEmpty() && levelOrderList.get(levelOrderList.size() - 1) == null) {
            levelOrderList.remove(levelOrderList.size() - 1);
        }

        return levelOrderList;
    }

    // Example usage
    public static void main(String[] args) {
        // Example input: [1,null,2,null,3,null,4,null,null]
        TreeNode root = new TreeNode(1);
        root.right = new TreeNode(2);
        root.right.right = new TreeNode(3);
        root.right.right.right = new TreeNode(4);

        BalanceaBinarySearchTree solution = new BalanceaBinarySearchTree();
        TreeNode balancedRoot = solution.balanceBST(root);
        List<Integer> levelOrder = solution.bstToLevelOrder(balancedRoot);

        System.out.println(levelOrder); // Output: [2, 1, 3, null, null, null, 4] or [3, 1, 4, null, 2] or any valid balanced BST
    }
}

File: python/BalanceaBinarySearchTree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


def balanceBST(root: TreeNode):
    # Step 1: Perform in-order traversal to collect all node values into a sorted list
    def inorder_traversal(node):
        if not node:
            return []
        return inorder_traversal(node.left) + [node.val] + inorder_traversal(node.right)

    sorted_list = inorder_traversal(root)

    # Step 2: Construct a balanced BST from sorted list
    def construct_balanced_bst(nums):
        if not nums:
            return None
        mid = len(nums) // 2
        root = TreeNode(nums[mid])
        root.left = construct_balanced_bst(nums[:mid])
        root.right = construct_balanced_bst(nums[mid + 1:])
        return root

    balanced_root = construct_balanced_bst(sorted_list)

    # Step 3: Convert balanced BST to level-order list format with None for missing nodes
    def bst_to_level_order(node):
        if not node:
            return []
        queue = [node]
        result = []
        while queue:
            current = queue.pop(0)
            if current:
                result.append(current.val)
                queue.append(current.left)
                queue.append(current.right)
            else:
                result.append(None)
        # Remove trailing None values
        while result and result[-1] is None:
            result.pop()
        return result

    return bst_to_level_order(balanced_root)


# Example usage:
# Input: root = [1,null,2,null,3,null,4,null,null]
root = TreeNode(1)
root.right = TreeNode(2)
root.right.right = TreeNode(3)
root.right.right.right = TreeNode(4)

output = balanceBST(root)
print(
    output)  # Output: [2, 1, 3, None, None, None, 4] or [3, 1, 4, None, 2] or any other balanced BST in level-order format
